internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -214,13 +214,26 @@ func (gsb *Balancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubC
 	} else if gsb.balancerPending != nil && gsb.balancerPending.subconns[sc] {
 		balToUpdate = gsb.balancerPending
 	}
-	gsb.mu.Unlock()
 	if balToUpdate == nil {
 		// SubConn belonged to a stale lb policy that has not yet fully closed,
 		// or the balancer was already closed.
+		gsb.mu.Unlock()
 		return
 	}
-	balToUpdate.UpdateSubConnState(sc, state)
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(balToUpdate.subconns, sc)
+	}
+	gsb.mu.Unlock()
+	if cb != nil {
+		cb(state)
```

Comment:
Would it be possible to handle this case first before grabbing any of the locks? Same comment applies to other places where we have the same pattern.

Reply:
We still need to do that other work, though (for now, until the old `UpdateSubConnState` is deleted).  Why do it first?  Once we delete `UpdateSubConnState` and require children clean up after themselves, this code will get a loooot simpler, as we can just pass everything through.

Reply:
Why do it first?

I thought something like this might work?
```golang
	if cb != nil {
		cb(state)
		return
	}
	// the existing code without any modification
```

Reply:
Not in this particular case, because there  is bookkeeping above that we must have for now.  Right now LB policies don't clean up after their own `SubConns`, and graceful switch is tracking them for that purpose.  When PF becomes the universal leaf policy, only PF will need to worry about that, and we can delete this interception stuff _entirely_ (YAY!).